### ISSUE TITLE
Admin category tree move doesn't recreate subcategory URL rewrites

### ISSUE THREAD

**Issue by VesnaS** [magento/magento2#12509]
### Preconditions
Magento version 2.1.6 (clean version with sample data)
PHP version 7.0.25
MySQL 5.7.20

### Steps to reproduce
1. Create Parent, Child and Grandchild categories

![parent_child_grandchild](https://user-images.githubusercontent.com/33281361/33481991-39d34bbe-d697-11e7-9274-f40b6b1fe535.png)
2. Magento will create correct URL rewrites for all three categories 
3. Move Child category to the same level as Parent category

![child_grandchild](https://user-images.githubusercontent.com/33281361/33482023-6bd81284-d697-11e7-87c5-a1d72a0cdf76.png)

### Expected result
1. Magento should create correct URL rewrites for Child category, and all of it's subcategories

### Actual result
1. Magento will recreate correct URL rewrite for Child category, but rewrite for Grandchild stays as it was

![new_grandchild](https://user-images.githubusercontent.com/33281361/33482222-47b0ec18-d698-11e7-9711-4c82e5e2f094.png)
2. Grandchild rewrite gets corrected only after category save

**Comment by magento-engcom-team** [comment]
Hello @VesnaS. Thanks for reporting. Seems we are not able to reproduce Your issue. Here we have attached the video with how we did it. So please, watch this video and tell us what we've been doing differently.
[urlrew.zip](https://github.com/magento/magento2/files/1521969/urlrew.zip)

**Comment by VesnaS** [comment]
Hello,
I watched your video, and you've repeated all the steps I've listed.
Here is the recording of me doing the same steps, with a different result from yours.
[reproduce.zip](https://github.com/magento/magento2/files/1522158/reproduce.zip)

**Comment by magento-engcom-team** [comment]
And would You @VesnaS be able to repeat all the steps in Your video, but with the system messages (there up on the screen) resolved, so that we can be sure that there are no unresolved issues that cause our issue. Thank You.

**Comment by VesnaS** [comment]
Hello,
I finally realized what the difference is between your and mine environment.
You have only one website, and I have multi-website configuration.
In order to be able to reproduce this bug, please:

-  Add one more website, store and store view

![test_website](https://user-images.githubusercontent.com/33281361/33555926-7610e308-d902-11e7-9caa-0a160e6e3aa5.png)

- Associate Default root category to the newly created store

- Repeat above mentioned steps and hopefully you will be able to see the problem I am talking about

**Comment by magento-engcom-team** [comment]
@VesnaS, thank you for your report.
We've created internal ticket(s) MAGETWO-85022 to track progress on the issue.

**Comment by novakivskiy** [comment]
#mageconf

**Comment by magento-engcom-team** [comment]
@novakivskiy thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by Bobstar040** [comment]
Hi Is there already a fix for this issue?

**Comment by novakivskiy** [comment]
Hi Magento Team, 
@magento-engcom-team  

Maybe I wrote a lot and it's not clear if you have any questions, I'll answer in slack


### Problem
In the admin area, if the executable URL does not contain a parameter in the URL request `___store`
or the executable code does not specify the store_code (       $this->getStoreManager()->setCurrentStore($store->getCode())). 
String  `$this->storeManager->getStore()->getId()` returns  store_code from the cookie frontend ([code](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Store/Model/StoreResolver.php#L91))

### The problem for this task
The problem is that when recursively loading the collections of categories, item collection  returned wrong store_id ($category->getStoreId()). This causes the value to be stored incorrectly in  DB. Example screen
![screen_1](https://user-images.githubusercontent.com/971021/35502518-f3ffddf0-04e5-11e8-8b96-d3b003c8fae3.png)
:

If category model doesn't contain data store_id, then the value is returned  `(int)$this->_storeManager->getStore()->getId();` [Code](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Model/Category.php#L578)

### The solution for this problem

In my opinion, the problem is global. The solution must be  return value  `0` (admin default store id) for code $this->_storeManager->getStore()->getId() in admin area. For example, create a individual store manager admin area.
Like is for url manager https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Backend/etc/adminhtml/di.xml#L12. Or fix the code in an existing class . This change is very sensitive to backward compatibility. I will not create a  pull request.

If the problem is solved locally, the solution looks like this:
1)  open Magento\Catalog\Controller\Adminhtml\Category\Move::execute(), 
2) create copy/past from  [Magento\Catalog\Controller\Adminhtml\Category\Edit::execute()](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Controller/Adminhtml/Category/Edit.php#L51)
  
`      $storeId = (int)$this->getRequest()->getParam('store');
        $store = $this->getStoreManager()->getStore($storeId);
        $this->getStoreManager()->setCurrentStore($store->getCode());`

to [Magento\Catalog\Controller\Adminhtml\Category\Move::execute()](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/Controller/Adminhtml/Category/Move.php#L51) similarly and copy past `private function getStoreManager()`

**Comment by juharintanen** [comment]
Can confirm the problem and @novakivskiy suggested modification did cure the problem! Thank you!

**Comment by lennartbee** [comment]
I can also confirm @novakivskiy solution worked.

**Comment by swnsma** [comment]
Fix from @novakivskiy solve the issue with url rewrites.
In other hand, flat tables still contains wrong **url_path** for stores with not '1' id

**Comment by nikunjskd20** [comment]
Didn't found this issue in 2.2 and 2.3 develop branch.

**Comment by engcom-Echo** [comment]
Hello All

I am not able to reproduce this issue on the 2.4-develop branch by provided steps.

**Testing scenario:**

**Steps to reproduce:**
1. Create Parent, Child and Grandchild categories
2. Magento will create correct URL rewrites for all three categories 
3. Move Child category to the same level as Parent category

**Result**
Magento creates correct URL rewrites for the Child category and all its subcategories
![wwww](https://user-images.githubusercontent.com/51681435/71480075-05c30780-2800-11ea-85de-4178fb662100.jpg)
Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.4-develop` branch. Thank you for collaboration.
